### PR TITLE
[SPARK-5950][SQL]Insert array into a metastore table saved as parquet should work when using datasource api

### PR DESCRIPTION
**Pull request by yhuai**
This PR contains the following changes:
1. Add a new method, `DataType.equalsIgnoreCompatibleNullability`, which is the middle ground between DataType's equality check and `DataType.equalsIgnoreNullability`. For two data types `from` and `to`, it does `equalsIgnoreNullability` as well as if the nullability of `from` is compatible with that of `to`. For example, the nullability of `ArrayType(IntegerType, containsNull = false)` is compatible with that of `ArrayType(IntegerType, containsNull = true)` (for an array without null values, we can always say it may contain null values). However,  the nullability of `ArrayType(IntegerType, containsNull = true)` is incompatible with that of `ArrayType(IntegerType, containsNull = false)` (for an array that may have null values, we cannot say it does not have null values).
2. For the `resolved` field of `InsertIntoTable`, use `equalsIgnoreCompatibleNullability` to replace the equality check of the data types.
3. For our data source write path, when appending data, we always use the schema of existing table to write the data. This is important for parquet, since nullability direct impacts the way to encode/decode values. If we do not do this, we may see corrupted values when reading values from a set of parquet files generated with different nullability settings.
4. When generating a new parquet table, we always set nullable/containsNull/valueContainsNull to true. So, we will not face situations that we cannot append data because containsNull/valueContainsNull in an Array/Map column of the existing table has already been set to `false`. This change makes the whole data pipeline more robust.
5. Update the equality check of JSON relation. Since JSON does not really cares nullability,  `equalsIgnoreNullability` seems a better choice to compare schemata from to JSON tables.

JIRA: https://issues.apache.org/jira/browse/SPARK-5950

Thanks @viirya for the initial work in #4729. 

cc @marmbrus @liancheng

### CODE REVIEW

**Comment by viirya** [comment]
I wouldn't like to say this. But, @liancheng, @yhuai I think you should more respect other's contribution...

You in #4782 made changes to `ParquetConversions`. That is almost as same as what I did in #4729.

Again, the point 4 in this pr:

> 4. When generating a new parquet table, we always set nullable/containsNull/valueContainsNull to true. So, we will not face situations that we cannot append data because containsNull/valueContainsNull in an Array/Map column of the existing table has already been set to false. This change makes the whole data pipeline more robust.

This is exactly what I did in #4729. I remember you said this is a bad idea and we would not use it.

Basically, I can't see why #4729 can't be modified and merged to do that. And I think it is important to respect other's contribution and the time spent on it...

**Comment by viirya** [comment]
@yhuai @liancheng I have no much to say. As you already opened these new prs that do what I submitted before in #4729, I closed the pr

Although you have explained why you did this, looks like you are not doing what you meant. You said you would not use my proposed approach, but then you create this to use it.

Maybe you should read these articles for pr etiquette.

http://readwrite.com/2014/07/02/github-pull-request-etiquette

http://www.quora.com/GitHub/Is-it-bad-etiquette-to-change-someones-pull-request-before-committing

File: mllib/src/main/scala/org/apache/spark/mllib/util/modelSaveLoad.scala [hunk 1]
```diff
@@ -110,7 +110,7 @@ private[mllib] object Loader {
       assert(loadedFields.contains(field.name), s"Unable to parse model data." +
         s"  Expected field with name ${field.name} was missing in loaded schema:" +
         s" ${loadedFields.mkString(", ")}")
-      assert(loadedFields(field.name) == field.dataType,
+      assert(DataType.equalsIgnoreNullability(loadedFields(field.name), field.dataType),
```
Comment (yhuai) [inline mllib/src/main/scala/org/apache/spark/mllib/util/modelSaveLoad.scala]: @mengxr Is it OK to ignore the nullability of ArrayType (containsNull field) and MapType (valueContainsNull field) in this check?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala [hunk 2]
```diff
@@ -172,9 +173,14 @@ private[sql] object ParquetRelation {
...
-    ParquetTypesConverter.writeMetaData(attributes, path, conf)
+    // This is a hack. We always set nullable/containsNull/valueContainsNull to true
+    // for the schema of a parquet data.
+    val schema =
+      DataType.alwaysNullable(StructType.fromAttributes(attributes)).asInstanceOf[StructType]
+    val newAttributes = schema.toAttributes
+    ParquetTypesConverter.writeMetaData(newAttributes, path, conf)
     new ParquetRelation(path.toString, Some(conf), sqlContext) {
-      override val output = attributes
+      override val output = newAttributes
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: @liancheng @marmbrus I am also changing the nullability for our old parquet write path to make the behavior consistent with our new write path. Let me know if there is any potential compatibility issue and we should revert this change.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 3]
```diff
@@ -592,13 +595,79 @@ class MetastoreDataSourcesSuite extends QueryTest with BeforeAndAfterEach {
...
+  }
+
   test("SPARK-6024 wide schema support") {
     // We will need 80 splits for this schema if the threshold is 4000.
     val schema = StructType((1 to 5000).map(i => StructField(s"c_${i}", StringType, true)))
     assert(
       schema.json.size > conf.schemaStringLengthThreshold,
       "To correctly test the fix of SPARK-6024, the value of " +
-      s"spark.sql.sources.schemaStringLengthThreshold needs to be less than ${schema.json.size}")
+        s"spark.sql.sources.schemaStringLengthThreshold needs to be less than ${schema.json.size}")
```
Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: looks like a spurious change.

**Comment by liancheng** [comment]
A typo in the PR description:

> However, the nullability of `ArrayType(IntegerType, containsNull = true)` is compatible with that of `ArrayType(IntegerType, containsNull = false)`

Here "compatible" should be "incompatible".

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 4]
```diff
@@ -198,6 +198,57 @@ object DataType {
...
     }
   }
+
+  /**
+   * Compares two types, ignoring compatible nullability of ArrayType, MapType, StructType.
+   *
+   * Compatible nullability is defined as follows:
+   *   - If `from` and `to` are ArrayTypes, `from` has a compatible nullability with `to`
+   *   if and only if `from.containsNull` is false, or both of `from.containsNull` and
+   *   `to.containsNull` are true.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Might be better to reword this definition to make it consistent with the code:

> If `from` and `to` are `ArrayType`s, `from` has a compatible nullablity with `to` if and only if `to.containsNull` is true or both `to.containsNull` and `from.containsNull` are false.

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Same applies to `MapType` and `StructType` below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 5]
```diff
@@ -198,6 +198,57 @@ object DataType {
...
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+
+  /** Sets all nullable/containsNull/valueContainsNull to true. */
+  private[spark] def alwaysNullable(dataType: DataType): DataType = dataType match {
+    case ArrayType(elementType, _) =>
+      ArrayType(alwaysNullable(elementType), containsNull = true)
+    case MapType(keyType, valueType, _) =>
+      MapType(alwaysNullable(keyType), alwaysNullable(valueType), true)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Nit: Use named parameter for `valueContainsNull`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DataTypeSuite.scala [hunk 6]
```diff
@@ -115,4 +115,84 @@ class DataTypeSuite extends FunSuite {
...
+    to = MapType(StringType, DoubleType, valueContainsNull = true),
+    expected = true)
+  checkEqualsIgnoreCompatibleNullability(
+    from = MapType(StringType, DoubleType, valueContainsNull = true),
+    to = MapType(StringType, DoubleType, valueContainsNull = false),
+    expected = false)
+  checkEqualsIgnoreCompatibleNullability(
+    from = MapType(StringType, ArrayType(IntegerType, true), valueContainsNull = true),
+    to = MapType(StringType,  ArrayType(IntegerType, false), valueContainsNull = true),
+    expected = false)
```
Comment (liancheng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/DataTypeSuite.scala]: Would be good to add another test case to show nested case:

```scala
  checkEqualsIgnoreCompatibleNullability(
    from = MapType(StringType, ArrayType(IntegerType, false), valueContainsNull = true),
    to = MapType(StringType,  ArrayType(IntegerType, true), valueContainsNull = true),
    expected = true)
```

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala (hunk 2 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: Should we also make data types of `ParquetRelation.output` always nullable?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: Never mind, since we always write nullable data, it should be OK to leave `ParquetRelation.output` untouched.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: Verified that when merging schemas, official Parquet implementation will handle nullability (repetition level) properly. So our change should be safe for interoperation with other systems that support Parquet schema evolving.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala [hunk 7]
```diff
@@ -172,9 +173,14 @@ private[sql] object ParquetRelation {
       sqlContext.conf.parquetCompressionCodec.toUpperCase, CompressionCodecName.UNCOMPRESSED)
       .name())
     ParquetRelation.enableLogForwarding()
-    ParquetTypesConverter.writeMetaData(attributes, path, conf)
+    // This is a hack. We always set nullable/containsNull/valueContainsNull to true
+    // for the schema of a parquet data.
+    val schema =
+      DataType.alwaysNullable(StructType.fromAttributes(attributes)).asInstanceOf[StructType]
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: Should also make `ParquetRelation2.schema` always nullable.

File: sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala [hunk 8]
```diff
@@ -131,7 +131,7 @@ private[sql] case class JSONRelation(
 
   override def equals(other: Any): Boolean = other match {
     case that: JSONRelation =>
-      (this.path == that.path) && (this.schema == that.schema)
+      (this.path == that.path) && (DataType.equalsIgnoreNullability(this.schema, that.schema))
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala]: Should also have this in `ParquetRelation2.equals`.

**Comment by liancheng** [comment]
@yhuai @marmbrus This PR basically makes both JSON and Parquet data sources always nullable. I agree that this makes most common cases more robust. Should we take a more aggressive step and just remove `nullable`/`containsNull`/`valueContainsNull` fields and assume all data sources are always nullable? Up until now, I can only see troubles rather than any benefits brought by nullability. Of course, this is out of the scope of this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala (hunk 7 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetRelation.scala]: I am not sure if we want to do this. For `createEmpty` (this change), we are creating a new parquet dataset. But, for ParquetRelation2, the schema may be created for an existing dataset. For every place we create a new parquet dataset, we will call alwaysNullable.

**Comment by marmbrus** [comment]
@liancheng I said basically the same thing to @yhuai when we were discussing it last week.  It does seem like nullability has been only a burden.  That said, it could allow us to do quite a few optimizations in the future if we keep it, so I'm kind of torn.  We should discuss and maybe benchmark.

Stylistically, I think we should be adding these methods to datatypes themselves.  I'd suggest that `DataType`s have a method called `sameType` that returns true when types are the same modulo nullability, instead of having a static method in `DataType`.  Pretty much every place where we now uses `==` should be using this function instead.  I'd also have  method `asNullable` that returns a `DataType` with nullability flipped on.  This does not have to happen in this PR if its too much work.

File: sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala (hunk 8 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala]: Since nullability affects how parquet encode and decode data, I think it is better to leave `ParquetRelation2.equals` unchanged for now.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala]: Please ignore my comment. Will change it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 9]
```diff
@@ -198,6 +198,43 @@ object DataType {
...
+   *   if and only if `to.containsNull` is true, or both of `from.containsNull` and
+   *   `to.containsNull` are false.
+   *   - If `from` and `to` are MapTypes, `from` has a compatible nullability with `to`
+   *   if and only if `to.valueContainsNull` is true, or both of `from.valueContainsNull` and
+   *   `to.valueContainsNull` are false.
+   *   - If `from` and `to` are StructTypes, `from` has a compatible nullability with `to`
+   *   if and only if for all every pair of fields, `to.nullable` is true, or both
+   *   of `fromField.nullable` and `toField.nullable` are false.
+   */
+  private[sql] def equalsIgnoreCompatibleNullability(from: DataType, to: DataType): Boolean = {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: We can introduce a method to the class of `DataType` based on this one later (I am not sure what will be a good name. I thought about `compatibleWith`, but I feel it is not very accurate).

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DataTypeSuite.scala (hunk 6 above)
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/types/DataTypeSuite.scala]: Done

**Comment by yhuai** [comment]
test this please
